### DISCUSSION TITLE
Additional resources on alignment

### DISCUSSION THREAD

**Discussion by AlecGoncharow** [sotrh/learn-wgpu#296] [Ideas]
I noticed you have a [page](https://sotrh.github.io/learn-wgpu/showcase/alignment/#alignment-of-vertex-and-index-buffers) on alignment and just wanted to drop a few links about aligned offset in uniform buffers as I have run into this in my own project. 

A great resource for GLSL, which appears to be tackling similar topics is the [Uniform block layout section on the Advanced GLSL page on Learn OpenGL](https://learnopengl.com/Advanced-OpenGL/Advanced-GLSL). 
This includes this example:
```glsl

layout (std140) uniform ExampleBlock
{
                     // base alignment  // aligned offset
    float value;     // 4               // 0 
    vec3 vector;     // 16              // 16  (offset must be multiple of 16 so 4->16)
    mat4 matrix;     // 16              // 32  (column 0)
                     // 16              // 48  (column 1)
                     // 16              // 64  (column 2)
                     // 16              // 80  (column 3)
    float values[3]; // 16              // 96  (values[0])
                     // 16              // 112 (values[1])
                     // 16              // 128 (values[2])
    bool boolean;    // 4               // 144
    int integer;     // 4               // 148
}; 
```
Which I found particularly helpful. One additional thing to note is that `mat4` (and likely `mat3` but I haven't tested it) is also required to be on a multiple of 16 aligned offset. 

Another particularly sharp edge is that the `mat3` type in GLSL also assumes each column has the same alignment as a `vec4` type, not sure if this is true in WGSL as well. 

FWIW it seems this is such an annoying thing to deal with that [some recommend not using `vec3`'s at all](https://stackoverflow.com/questions/38172696/should-i-ever-use-a-vec3-inside-of-a-uniform-buffer-or-shader-storage-buffer-o), I am not sold on this idea, but I believe it is worth acknowledging.

**Comment by sotrh** [comment]
Hmm interesting. Float arrays really get padded like that? You'd think they'd just have padding on the end.
